src/vs/editor/contrib/linesOperations/linesOperations.ts

```diff
@@ -112,18 +112,74 @@ abstract class AbstractMoveLinesAction extends EditorAction {
 		let commands: ICommand[] = [];
 		let selections = editor.getSelections() || [];
 		let autoIndent = editor.getConfiguration().autoIndent;
+		let movingMultipleLines = false;
+		let selectionDirectionDown = true;
+		let newSelections: Selection[] = [];
+
+		if (selections.length > 1) {
+			movingMultipleLines = selections[0].endLineNumber !== selections[selections.length - 1].endLineNumber;
+
+			if (movingMultipleLines) {
+				// Stash selections while processing, set new selections +/- line-change only
+				for (const selection of selections) {
+					let startCol = selection.getStartPosition().column;
+					let startLine = selection.getStartPosition().lineNumber;
+					let newStartLine = this.down ? startLine + 1 : startLine - 1;
+
+					let endCol = selection.getEndPosition().column;
+					let endLine = selection.getEndPosition().lineNumber;
+					let newEndLine = this.down ? endLine + 1 : endLine - 1;
+
+					newSelections.push(new Selection(newStartLine, startCol, newEndLine, endCol));
+				}
+
+				// Work only with one selection per line
+				selections = selections.filter((s, idx, arr) => {
+					return arr.map(sel => sel['endLineNumber']).indexOf(s['endLineNumber']) === idx;
+				});
+
+				selectionDirectionDown = selections[0].endLineNumber < selections[selections.length - 1].endLineNumber;
+
+				editor.pushUndoStop();
+
+				if (selectionDirectionDown === this.down) {
+					selections.reverse();
+				}
+			}
+		}
 
 		for (const selection of selections) {
-			commands.push(new MoveLinesCommand(selection, this.down, autoIndent));
+			if (movingMultipleLines) {
+				// If we are at the beginning/end of document and multiple lines are moved, abort.
+				let model = editor.getModel();
+				let lineCount = model !== null ? model.getLineCount() : 0;
+
+				if ((selection.startLineNumber === 1 && !this.down) || (selection.endLineNumber === lineCount && this.down)) {
+					editor.pushUndoStop();
+					return;
+				}
+
+				editor.executeCommand(this.id, new MoveLinesCommand(selection, this.down, autoIndent));
+			} else {
+				commands.push(new MoveLinesCommand(selection, this.down, autoIndent));
+			}
 		}
 
-		editor.pushUndoStop();
-		editor.executeCommands(this.id, commands);
-		editor.pushUndoStop();
+		if (movingMultipleLines) {
+			if (newSelections.length > 1) {
+				editor.setSelections(newSelections);
+			}
+
+			editor.pushUndoStop();
+		} else {
+			editor.pushUndoStop();
```

Comment:
Could we unify this part of the code between the two cases? We could construct the commands array entirely first and then call `executeCommands` on this `commands` array. Then after this, if we have moved multiple lines, we can set the selections.

Reply:
This didn't work out for me. For some reason it just doesn't process all of the commands